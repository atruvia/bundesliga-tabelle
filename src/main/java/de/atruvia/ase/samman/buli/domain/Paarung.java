package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.BEENDET;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.GEPLANT;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.LAUFEND;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;

import java.util.function.Function;

import org.jmolecules.ddd.annotation.ValueObject;

import lombok.Builder;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import lombok.With;
import lombok.experimental.Accessors;

@Value
@Builder(toBuilder = true)
@Accessors(fluent = true)
@ValueObject
public class Paarung {

	@Value
	public class PaarungView {

		ViewDirection direction;
		Function<Paarung, Entry> self;
		Function<Paarung, Entry> gegner;
		@Getter(lazy = true)
		Ergebnis ergebnis = calcErgebnis();

		public Entry self() {
			return self.apply(paarung());
		}

		public Entry gegner() {
			return gegner.apply(paarung());
		}

		public int tore() {
			return self().tore();
		}

		public int gegentore() {
			return gegner().tore();
		}

		public boolean isGeplant() {
			return paarung().isGeplant();
		}

		public boolean isLaufend() {
			return paarung().isLaufend();
		}

		private Ergebnis calcErgebnis() {
			var tore = tore();
			var gegentore = gegentore();
			if (tore == gegentore) {
				return UNENTSCHIEDEN;
			} else if (tore > gegentore) {
				return SIEG;
			} else {
				return NIEDERLAGE;
			}
		}

		public ErgebnisTyp ergebnisTyp() {
			return paarung().ergebnisTyp;
		}

		private Paarung paarung() {
			return Paarung.this;
		}

	}

	public enum ViewDirection {
		HEIM, AUSWAERTS
	}

	@RequiredArgsConstructor
	@Getter
	@Accessors(fluent = true)
	public enum Ergebnis {

		SIEG, UNENTSCHIEDEN, NIEDERLAGE;

		public int punkte() {
			return switch (this) {
			case SIEG -> 3;
			case UNENTSCHIEDEN -> 1;
			case NIEDERLAGE -> 0;
			};
		}

	}

	public enum ErgebnisTyp {
		GEPLANT, LAUFEND, BEENDET
	}

	@Value
	@Builder(toBuilder = true)
	public static class Entry {
		Team team;
		@With
		int tore;

		public Entry(Team team, int tore) {
			if (tore < 0) {
				throw new IllegalArgumentException("Tore darf nicht negativ sein, ist aber " + tore);
			}
			this.team = team;
			this.tore = tore;
		}

	}

	@Builder.Default
	ErgebnisTyp ergebnisTyp = GEPLANT;

	Entry heim, gast;

	public boolean isGeplant() {
		return ergebnisTypIs(GEPLANT);
	}

	public boolean isLaufend() {
		return ergebnisTypIs(LAUFEND);
	}

	private boolean ergebnisTypIs(ErgebnisTyp ergebnisTyp) {
		return this.ergebnisTyp == ergebnisTyp;
	}

	public Paarung withErgebnis(int toreHeim, int toreGast) {
		return toBuilder().endergebnis(toreHeim, toreGast).build();
	}

	public ViewDirection viewDirection() {
		return HEIM;
	}

	public static class PaarungBuilder {

		public static PaarungBuilder paarung(String heim, String gast) {
			return paarung(team(heim), team(gast));
		}

		private static Team team(String name) {
			return Team.builder().name(name).build();
		}

		public static PaarungBuilder paarung(Team heim, Team gast) {
			return Paarung.builder().heim(entry(heim)).gast(entry(gast));
		}

		public PaarungBuilder endergebnis(int toreHeim, int toreGast) {
			return ergebnis(BEENDET, toreHeim, toreGast);
		}

		public PaarungBuilder zwischenergebnis(int toreHeim, int toreGast) {
			return ergebnis(LAUFEND, toreHeim, toreGast);
		}

		PaarungBuilder ergebnis(ErgebnisTyp ergebnisTyp, int toreHeim, int toreGast) {
			return ergebnisTyp(ergebnisTyp).goals(toreHeim, toreGast);
		}

		private static Entry entry(Team team) {
			return Entry.builder().team(team).build();
		}

		public PaarungBuilder goals(int toreHeim, int toreGast) {
			return heim(heim.withTore(toreHeim)).gast(gast.withTore(toreGast));
		}

	}

	public PaarungView viewForTeam(ViewDirection viewDirection) {
		return switch (viewDirection) {
		case HEIM -> new PaarungView(HEIM, Paarung::heim, Paarung::gast);
		case AUSWAERTS -> new PaarungView(AUSWAERTS, Paarung::gast, Paarung::heim);
		};
	}

}