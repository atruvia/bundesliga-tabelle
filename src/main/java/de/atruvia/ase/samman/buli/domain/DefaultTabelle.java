package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;
import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;
import static java.util.function.Function.identity;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toMap;
import static java.util.stream.Stream.concat;
import static lombok.AccessLevel.PRIVATE;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;

import org.jmolecules.ddd.annotation.ValueObject;

import de.atruvia.ase.samman.buli.domain.Paarung.PaarungView;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import lombok.experimental.Accessors;

@RequiredArgsConstructor
@ValueObject
public class DefaultTabelle implements Tabelle {

	@Value
	@Accessors(fluent = true)
	private static class OrdnungsElement implements Comparable<OrdnungsElement> {

		private static final Comparator<OrdnungsElement> comparator = comparing(value(TabellenPlatz::punkte)) //
				.thenComparing(value(TabellenPlatz::torDifferenz)) //
				.thenComparing(value(TabellenPlatz::gesamtTore)) //
				.thenComparing(direkterVergleichGesamt()) //
				.thenComparing(direkterVergleichAuswaertsTore()) //
				.thenComparing(value(TabellenPlatz::auswaertsTore)) //
				.reversed();

		private static Comparator<OrdnungsElement> direkterVergleichGesamt() {
			return compareWithSwapped((that, other) -> whereToreIs(that.tabellenPlatz, gegnerIs(other.team())));
		}

		private static Comparator<OrdnungsElement> direkterVergleichAuswaertsTore() {
			return compareWithSwapped(
					(that, other) -> whereToreIs(that.tabellenPlatz, gegnerIs(other.team()).and(isAuswaerts())));
		}

		private static <T, R extends Comparable<R>> Comparator<T> compareWithSwapped(BiFunction<T, T, R> biFunction) {
			return (o1, o2) -> biFunction.apply(o1, o2).compareTo(biFunction.apply(o2, o1));
		}

		private Team team() {
			return tabellenPlatz.team();
		}

		private static Predicate<PaarungView> gegnerIs(Team gegner) {
			return e -> Objects.equals(e.gegner().team().id(), gegner.id());
		}

		private static Predicate<PaarungView> isAuswaerts() {
			return e -> Objects.equals(e.direction(), AUSWAERTS);
		}

		private static int whereToreIs(TabellenPlatz tabellenPlatz, Predicate<PaarungView> filter) {
			return tabellenPlatz.paarungenStream().filter(filter).map(PaarungView::tore).mapToInt(Integer::valueOf)
					.sum();
		}

		@Getter(value = PRIVATE)
		TabellenPlatz tabellenPlatz;

		@Override
		public int hashCode() {
			return 0;
		}

		private static <T> Function<OrdnungsElement, T> value(Function<TabellenPlatz, T> function) {
			return t -> function.apply(t.tabellenPlatz);
		}

		@Override
		public boolean equals(Object o) {
			return comparator.compare(this, (OrdnungsElement) o) == 0;
		}

		@Override
		public int compareTo(OrdnungsElement other) {
			return comparator.thenComparing(e -> e.team().name()).compare(this, other);
		}

	}

	private final Collection<TabellenPlatz> entries;

	public DefaultTabelle() {
		this(emptyList());
	}

	@Override
	public Tabelle add(Paarung paarung) {
		var existing = entries().stream();
		var toAdd = Stream.of(newEntry(paarung.viewForTeam(HEIM)), newEntry(paarung.viewForTeam(AUSWAERTS)));
		var newEntries = concat(existing, toAdd) //
				.collect(toMap(t -> t.team().id(), identity(), TabellenPlatz::mergeWith)) //
				.values();
		return new DefaultTabelle(newEntries);
	}

	private static TabellenPlatz newEntry(PaarungView paarung) {
		var builder = TabellenPlatz.builder().team(paarung.self().team());
		return (paarung.isGeplant() ? builder : builder.paarung(paarung)).build();
	}

	public List<TabellenPlatz> entries() {
		// TODO make it side-effect-free
		AtomicInteger counter = new AtomicInteger(1);
		Map<OrdnungsElement, List<TabellenPlatz>> platzGruppen = entries.stream()
				.collect(groupingBy(OrdnungsElement::new));
		return new TreeMap<>(platzGruppen).values().stream() //
				.flatMap(t -> makeGroup(counter, t)) //
				.toList();
	}

	private static Stream<TabellenPlatz> makeGroup(AtomicInteger counter, List<TabellenPlatz> tabellenPlaetze) {
		Rank rank = new Rank(counter.getAndAdd(tabellenPlaetze.size()));
		return tabellenPlaetze.stream() //
				.sorted(comparing(OrdnungsElement::new)) //
				.map(tp -> tp.withRank(rank));
	}

}
