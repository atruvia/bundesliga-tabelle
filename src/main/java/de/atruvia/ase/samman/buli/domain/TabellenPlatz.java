package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.BEENDET;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;
import static de.atruvia.ase.samman.buli.util.Merger.lastNonNull;
import static de.atruvia.ase.samman.buli.util.Merger.merge;
import static java.util.Arrays.asList;
import static java.util.stream.IntStream.range;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BinaryOperator;
import java.util.stream.Stream;

import de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis;
import de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp;
import de.atruvia.ase.samman.buli.domain.Paarung.PaarungView;
import de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection;
import de.atruvia.ase.samman.buli.util.Merger.Mergeable;
import lombok.Builder;
import lombok.Value;
import lombok.With;
import lombok.experimental.Accessors;

@Value
@Accessors(fluent = true)
@Builder
public class TabellenPlatz implements Mergeable<TabellenPlatz> {

	@Value
	private static class ErgebnisEntry {
		Ergebnis ergebnis;
		ErgebnisTyp ergebnisTyp;
	}

	private static final BinaryOperator<Integer> adder = (i1, i2) -> i1 + i2;

	URI wappen;
	@With
	int platz;
	@With
	String team;
	int spiele;
	List<ErgebnisEntry> ergebnisse;
	int punkte;
	Map<ViewDirection, Integer> tore;
	Map<ViewDirection, Integer> gegentore;
	PaarungView laufendesSpiel;

	public List<Ergebnis> ergebnisse() {
		return collectToList(ergebnisseStream());
	}

	public List<Ergebnis> ergebnisse(ErgebnisTyp... ergebnisTyp) {
		return collectToList(ergebnisseStream().filter(e -> entryErgebnisIsTypeOf(e, ergebnisTyp)));
	}

	private Stream<ErgebnisEntry> ergebnisseStream() {
		return ergebnisse.stream();
	}

	private static List<Ergebnis> collectToList(Stream<ErgebnisEntry> filter) {
		return filter.map(ErgebnisEntry::ergebnis).toList();
	}

	private static boolean entryErgebnisIsTypeOf(ErgebnisEntry e, ErgebnisTyp... ergebnisTyp) {
		return asList(ergebnisTyp).contains(e.ergebnisTyp());
	}

	public int gesamtTore() {
		return heimtore() + auswaertsTore();
	}

	public int gesamtGegentore() {
		return heimGegentore() + auswaertsGegentore();
	}

	public int heimtore() {
		return tore().getOrDefault(HEIM, 0);
	}

	public int auswaertsTore() {
		return tore().getOrDefault(AUSWAERTS, 0);
	}

	public int heimGegentore() {
		return gegentore().getOrDefault(HEIM, 0);
	}

	public int auswaertsGegentore() {
		return gegentore().getOrDefault(AUSWAERTS, 0);
	}

	public int torDifferenz() {
		return gesamtTore() - gesamtGegentore();
	}

	public TabellenPlatzBuilder toBuilder() {
		TabellenPlatzBuilder builder = new TabellenPlatzBuilder();
		builder.wappen = wappen;
		builder.platz = platz;
		builder.team = team;
		builder.spiele = spiele;
		builder.ergebnisse = new ArrayList<>(ergebnisse);
		builder.punkte = punkte;
		builder.tore = new HashMap<>(tore);
		builder.gegentore = new HashMap<>(gegentore);
		builder.laufendesSpiel = laufendesSpiel;
		return builder;
	}

	public static class TabellenPlatzBuilder {

		public TabellenPlatzBuilder() {
			ergebnisse = new ArrayList<>();
			tore = new HashMap<>();
			gegentore = new HashMap<>();
		}

		public TabellenPlatzBuilder ergebnis(Ergebnis ergebnis, ErgebnisTyp ergebnisTyp) {
			ergebnisse.add(new ErgebnisEntry(ergebnis, ergebnisTyp));
			return this;
		}

		public TabellenPlatzBuilder withGegentore(ViewDirection direction, int anzahl) {
			gegentore.put(direction, anzahl);
			return this;
		}

		public TabellenPlatzBuilder withTore(ViewDirection viewDirection, int anzahl) {
			tore.put(viewDirection, anzahl);
			return this;
		}

	}

	@Override
	public TabellenPlatz mergeWith(TabellenPlatz other) {
		return builder() //
				.team(lastNonNull(team, other.team)) //
				.ergebnisse(merge(ergebnisse, other.ergebnisse)) //
				.spiele(merge(spiele, other.spiele)) //
				.punkte(merge(punkte, other.punkte)) //
				.tore(merge(adder, tore, other.tore)) //
				.gegentore(merge(adder, gegentore, other.gegentore)) //
				.wappen(lastNonNull(wappen, other.wappen)) //
				.laufendesSpiel(lastNonNull(laufendesSpiel, other.laufendesSpiel)) //
				.build();
	}

	public int siege() {
		return countAnzahl(SIEG);
	}

	public int unentschieden() {
		return countAnzahl(UNENTSCHIEDEN);
	}

	public int niederlagen() {
		return countAnzahl(NIEDERLAGE);
	}

	private int countAnzahl(Ergebnis type) {
		return (int) ergebnisseStream().map(ErgebnisEntry::ergebnis).filter(type::equals).count();
	}

	public Ergebnis[] tendenz() {
		return last(5, ergebnisse(BEENDET)).toArray(Ergebnis[]::new);
	}

	private static <T> Stream<T> last(int count, List<T> elements) {
		int size = elements.size();
		return range(0, count).mapToObj(i -> {
			int index = size - i - 1;
			return index >= 0 ? elements.get(index) : null;
		});
	}

}