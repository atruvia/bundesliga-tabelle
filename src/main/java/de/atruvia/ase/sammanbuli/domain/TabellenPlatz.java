package de.atruvia.ase.sammanbuli.domain;

import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static java.util.Collections.emptyMap;
import static java.util.stream.Collectors.toMap;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.stream.Stream;

import de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis;
import lombok.Builder;
import lombok.Value;
import lombok.With;

@Value
@Builder
public class TabellenPlatz {

	static TabellenPlatz NULL = new TabellenPlatz(0, "", 0, emptyMap(), 0, 0, 0, 0, 0);

	@With
	int platz;
	@With
	String team;
	@Builder.Default
	int spiele = 1;
	Map<Ergebnis, Integer> ergebnisse;
	int punkte;
	int toreHeim;
	int toreAuswaerts;
	int gegentoreHeim;
	int gegentoreAuswaerts;

	public int getTore() {
		return toreHeim + toreAuswaerts;
	}

	public int getGegentore() {
		return gegentoreHeim + gegentoreAuswaerts;
	}

	static class TabellenPlatzBuilder {

		TabellenPlatzBuilder() {
			ergebnisse = new HashMap<>();
		}

		public TabellenPlatz.TabellenPlatzBuilder ergebnis(Ergebnis ergebnis) {
			ergebnisse.merge(ergebnis, 1, (a, b) -> a + b);
			return this;
		}

	}

	public int getTorDifferenz() {
		return getTore() - getGegentore();
	}

	public TabellenPlatz merge(TabellenPlatz other) {
		return TabellenPlatz.builder() //
				.ergebnisse(merge(this.ergebnisse, other.ergebnisse)) //
				.spiele(this.spiele + other.spiele) //
				.punkte(this.punkte + other.punkte) //
				.toreHeim(this.toreHeim + other.toreHeim) //
				.toreAuswaerts(this.toreAuswaerts + other.toreAuswaerts) //
				.gegentoreHeim(this.gegentoreHeim + other.gegentoreHeim) //
				.gegentoreAuswaerts(this.gegentoreAuswaerts + other.gegentoreAuswaerts) //
				.build();
	}

	private static Map<Ergebnis, Integer> merge(Map<Ergebnis, Integer> map1, Map<Ergebnis, Integer> map2) {
		return Stream.of(map1, map2).map(Map::entrySet).flatMap(Collection::stream)
				.collect(toMap(Entry::getKey, Entry::getValue, Integer::sum));
	}

	public int getGewonnen() {
		return ergebnis(SIEG);
	}

	public int getUnentschieden() {
		return ergebnis(UNENTSCHIEDEN);
	}

	public int getVerloren() {
		return ergebnis(NIEDERLAGE);
	}

	private int ergebnis(Ergebnis type) {
		return ergebnisse.getOrDefault(type, 0);
	}

}