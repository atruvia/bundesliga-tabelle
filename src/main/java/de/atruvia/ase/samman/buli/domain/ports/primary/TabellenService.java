package de.atruvia.ase.samman.buli.domain.ports.primary;

import java.util.List;

import de.atruvia.ase.samman.buli.domain.TabellenPlatz;

public interface TabellenService {

	public interface Tabelle {
		String name();

		List<TabellenPlatz> eintraege();
	}

	Tabelle erstelleTabelle(String league, String season);

}
