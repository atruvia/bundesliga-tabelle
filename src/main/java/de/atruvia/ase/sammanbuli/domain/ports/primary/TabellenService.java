package de.atruvia.ase.sammanbuli.domain.ports.primary;

import java.util.List;

import de.atruvia.ase.samman.buli.domain.TabellenPlatz;

public interface TabellenService {

	List<TabellenPlatz> erstelleTabelle(String league, String season);

}
