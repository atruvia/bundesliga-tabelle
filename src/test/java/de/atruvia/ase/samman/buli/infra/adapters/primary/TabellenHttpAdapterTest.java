package de.atruvia.ase.samman.buli.infra.adapters.primary;

import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static org.hamcrest.CoreMatchers.is;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.net.URI;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.web.servlet.MockMvc;

import de.atruvia.ase.samman.buli.domain.TabellenPlatz;
import de.atruvia.ase.samman.buli.domain.ports.primary.TabellenService;

@SpringBootTest
@AutoConfigureMockMvc
class TabellenHttpAdapterTest {

	@Autowired
	private MockMvc mockMvc;

	@MockBean
	private TabellenService tabellenService;

	@Test
	void shouldReturnDefaultMessage() throws Exception {
		String league = "bl1";
		String season = "2022";

		TabellenPlatz platz1 = TabellenPlatz.builder().wappen(URI.create("proto://wappen-team-10")).team("Team 10")
				.spiele(11).ergebnisse(Map.of(SIEG, 12, UNENTSCHIEDEN, 13, NIEDERLAGE, 14)).toreHeim(15)
				.toreAuswaerts(16).gegentoreHeim(17).gegentoreAuswaerts(18).punkte(19).build();
		TabellenPlatz platz2 = TabellenPlatz.builder().wappen(URI.create("proto://wappen-team-20")).team("Team 20")
				.spiele(21).ergebnisse(Map.of(SIEG, 22, UNENTSCHIEDEN, 23, NIEDERLAGE, 24)).toreHeim(25)
				.toreAuswaerts(26).gegentoreHeim(27).gegentoreAuswaerts(28).punkte(29).build();
		when(tabellenService.erstelleTabelle(league, season)).thenReturn(List.of(platz1, platz2));

		this.mockMvc.perform(get("/tabelle/" + league + "/" + season)) //
				.andDo(print()) //
				.andExpect(status().isOk()) //
				.andExpect(jsonPath("$.[0].team", is(platz1.getTeam()))) //
				.andExpect(jsonPath("$.[0].spiele", is(platz1.getSpiele()))) //
				.andExpect(jsonPath("$.[0].siege", is(platz1.getSiege()))) //
				.andExpect(jsonPath("$.[0].unentschieden", is(platz1.getUnentschieden()))) //
				.andExpect(jsonPath("$.[0].niederlagen", is(platz1.getNiederlagen()))) //
				.andExpect(jsonPath("$.[0].tore", is(platz1.getTore()))) //
				.andExpect(jsonPath("$.[0].gegentore", is(platz1.getGegentore()))) //
				.andExpect(jsonPath("$.[0].tordifferenz", is(platz1.getTorDifferenz()))) //
				.andExpect(jsonPath("$.[0].punkte", is(platz1.getPunkte()))) //
				//
				.andExpect(jsonPath("$.[1].team", is(platz2.getTeam()))) //
				.andExpect(jsonPath("$.[1].spiele", is(platz2.getSpiele()))) //
				.andExpect(jsonPath("$.[1].siege", is(platz2.getSiege()))) //
				.andExpect(jsonPath("$.[1].unentschieden", is(platz2.getUnentschieden()))) //
				.andExpect(jsonPath("$.[1].niederlagen", is(platz2.getNiederlagen()))) //
				.andExpect(jsonPath("$.[1].tore", is(platz2.getTore()))) //
				.andExpect(jsonPath("$.[1].gegentore", is(platz2.getGegentore()))) //
				.andExpect(jsonPath("$.[1].tordifferenz", is(platz2.getTorDifferenz()))) //
				.andExpect(jsonPath("$.[1].punkte", is(platz2.getPunkte()))) //
		;

	}

}
