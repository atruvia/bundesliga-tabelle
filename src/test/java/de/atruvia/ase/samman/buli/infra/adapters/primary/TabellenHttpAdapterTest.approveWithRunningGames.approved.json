[
  {
    "platz": 1,
    "wappen": "https://upload.wikimedia.org/wikipedia/de/b/b3/Fc_st_pauli_logo.svg",
    "team": "FC St. Pauli",
    "spiele": 26,
    "punkte": 54,
    "tore": 48,
    "gegentore": 25,
    "tordifferenz": 23,
    "siege": 15,
    "unentschieden": 9,
    "niederlagen": 2,
    "letzte5": "SSNSS"
  },
  {
    "platz": 2,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/thumb/3/30/Holstein_Kiel_Logo.svg/300px-Holstein_Kiel_Logo.svg.png",
    "team": "Holstein Kiel",
    "spiele": 27,
    "punkte": 52,
    "tore": 50,
    "gegentore": 34,
    "tordifferenz": 16,
    "siege": 16,
    "unentschieden": 4,
    "niederlagen": 7,
    "letzte5": "SSUNS",
    "laufendesSpiel": "1:0"
  },
  {
    "platz": 3,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/f/f7/Hamburger_SV_logo.svg",
    "team": "Hamburger SV",
    "spiele": 26,
    "punkte": 44,
    "tore": 50,
    "gegentore": 37,
    "tordifferenz": 13,
    "siege": 13,
    "unentschieden": 5,
    "niederlagen": 8,
    "letzte5": "SNNSU"
  },
  {
    "platz": 4,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/thumb/9/94/Fortuna_D%C3%BCsseldorf.svg/150px-Fortuna_D%C3%BCsseldorf.svg.png",
    "team": "Fortuna DÃ¼sseldorf",
    "spiele": 27,
    "punkte": 43,
    "tore": 56,
    "gegentore": 35,
    "tordifferenz": 21,
    "siege": 12,
    "unentschieden": 7,
    "niederlagen": 8,
    "letzte5": "SSUSU",
    "laufendesSpiel": "0:1"
  },
  {
    "platz": 5,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/c/cd/Hannover_96_Logo.svg",
    "team": "Hannover 96",
    "spiele": 26,
    "punkte": 40,
    "tore": 47,
    "gegentore": 35,
    "tordifferenz": 12,
    "siege": 10,
    "unentschieden": 10,
    "niederlagen": 6,
    "letzte5": "UUUNS"
  },
  {
    "platz": 6,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/e/e3/SC_Paderborn_07_Logo.svg",
    "team": "SC Paderborn 07",
    "spiele": 26,
    "punkte": 39,
    "tore": 40,
    "gegentore": 43,
    "tordifferenz": -3,
    "siege": 11,
    "unentschieden": 6,
    "niederlagen": 9,
    "letzte5": "NUUSN"
  },
  {
    "platz": 7,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/c/c8/Karlsruher_SC_Logo_2.svg",
    "team": "Karlsruher SC",
    "spiele": 26,
    "punkte": 38,
    "tore": 55,
    "gegentore": 41,
    "tordifferenz": 14,
    "siege": 10,
    "unentschieden": 8,
    "niederlagen": 8,
    "letzte5": "SNSSU"
  },
  {
    "platz": 8,
    "wappen": "https://i.imgur.com/pwuVbKr.png",
    "team": "SpVgg Greuther FÃ¼rth",
    "spiele": 26,
    "punkte": 38,
    "tore": 37,
    "gegentore": 38,
    "tordifferenz": -1,
    "siege": 11,
    "unentschieden": 5,
    "niederlagen": 10,
    "letzte5": "NNNSN"
  },
  {
    "platz": 9,
    "wappen": "https://i.imgur.com/apFwbYZ.png",
    "team": "Hertha BSC",
    "spiele": 26,
    "punkte": 37,
    "tore": 50,
    "gegentore": 43,
    "tordifferenz": 7,
    "siege": 10,
    "unentschieden": 7,
    "niederlagen": 9,
    "letzte5": "SNUUS"
  },
  {
    "platz": 10,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/d/d4/SV_Elversberg_Logo_2015.svg",
    "team": "SV Elversberg 07",
    "spiele": 27,
    "punkte": 36,
    "tore": 39,
    "gegentore": 43,
    "tordifferenz": -4,
    "siege": 10,
    "unentschieden": 6,
    "niederlagen": 11,
    "letzte5": "NSNNS",
    "laufendesSpiel": "0:0"
  },
  {
    "platz": 11,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/f/fa/1._FC_N%C3%BCrnberg_logo.svg",
    "team": "1. FC NÃ¼rnberg",
    "spiele": 26,
    "punkte": 36,
    "tore": 35,
    "gegentore": 45,
    "tordifferenz": -10,
    "siege": 10,
    "unentschieden": 6,
    "niederlagen": 10,
    "letzte5": "NSSNU"
  },
  {
    "platz": 12,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/d/d3/Logo_1_FC_Kaiserslautern.svg",
    "team": "1. FC Kaiserslautern",
    "spiele": 27,
    "punkte": 32,
    "tore": 43,
    "gegentore": 51,
    "tordifferenz": -8,
    "siege": 9,
    "unentschieden": 5,
    "niederlagen": 13,
    "letzte5": "USSNU",
    "laufendesSpiel": "1:0"
  },
  {
    "platz": 13,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/thumb/8/84/1._FC_Magdeburg.svg/210px-1._FC_Magdeburg.svg.png",
    "team": "1. FC Magdeburg",
    "spiele": 26,
    "punkte": 31,
    "tore": 38,
    "gegentore": 41,
    "tordifferenz": -3,
    "siege": 8,
    "unentschieden": 7,
    "niederlagen": 11,
    "letzte5": "NNUSN"
  },
  {
    "platz": 14,
    "wappen": "https://upload.wikimedia.org/wikipedia/de/thumb/3/3d/Logo_SV_Wehen_Wiesbaden.svg/1024px-Logo_SV_Wehen_Wiesbaden.svg.png",
    "team": "SV Wehen Wiesbaden",
    "spiele": 26,
    "punkte": 31,
    "tore": 30,
    "gegentore": 34,
    "tordifferenz": -4,
    "siege": 8,
    "unentschieden": 7,
    "niederlagen": 11,
    "letzte5": "NUSNN"
  },
  {
    "platz": 15,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/9/97/FC_Schalke_04_Logo.png",
    "team": "FC Schalke 04",
    "spiele": 26,
    "punkte": 30,
    "tore": 42,
    "gegentore": 54,
    "tordifferenz": -12,
    "siege": 9,
    "unentschieden": 3,
    "niederlagen": 14,
    "letzte5": "NUSNS"
  },
  {
    "platz": 16,
    "wappen": "https://upload.wikimedia.org/wikipedia/de/4/45/Logo_Eintracht_Braunschweig.svg",
    "team": "Eintracht Braunschweig",
    "spiele": 27,
    "punkte": 28,
    "tore": 25,
    "gegentore": 39,
    "tordifferenz": -14,
    "siege": 8,
    "unentschieden": 4,
    "niederlagen": 15,
    "letzte5": "SNNUN",
    "laufendesSpiel": "0:0"
  },
  {
    "platz": 17,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/0/00/FC_Hansa_Logo_seit_2009.png",
    "team": "FC Hansa Rostock",
    "spiele": 27,
    "punkte": 28,
    "tore": 24,
    "gegentore": 42,
    "tordifferenz": -18,
    "siege": 8,
    "unentschieden": 4,
    "niederlagen": 15,
    "letzte5": "SSNNU",
    "laufendesSpiel": "0:1"
  },
  {
    "platz": 18,
    "wappen": "https://upload.wikimedia.org/wikipedia/commons/2/24/Logo_Vfl_Osnabrueck_2017.png",
    "team": "VfL OsnabrÃ¼ck",
    "spiele": 26,
    "punkte": 18,
    "tore": 24,
    "gegentore": 53,
    "tordifferenz": -29,
    "siege": 3,
    "unentschieden": 9,
    "niederlagen": 14,
    "letzte5": "NNSSN"
  }
]
